File: 2021/10/07/large_group_positions.py
"""
https://leetcode.com/problems/positions-of-large-groups/
Author: Andrew Jarombek
Date: 10/7/2021
"""

from typing import List


def large_group_positions(s: str) -> List[List[int]]:
    result = []

    current = s[0]
    start = 0

    for i in range(1, len(s)):
        if s[i] != current:
            if i - start >= 3:
                result.append([start, i - 1])

            start = i
            current = s[i]

    if len(s) - start >= 3:
        result.append([start, len(s) - 1])

    return result

File: 2023/01/07/largeGroupPositions.js
/**
 * https://leetcode.com/problems/positions-of-large-groups/
 * @author Andrew Jarombek
 * @since 1/7/2023
 */

const largeGroupPositions = function(s) {
    const result = [];
    let prev = '';
    let start = 0;

    for (let i = 0; i < s.length; i++) {
        const c = s.charAt(i);

        if (c !== prev) {
            if (i > start + 2) {
                result.push([start, i-1]);
            }

            prev = c;
            start = i;
        }
    }

    if (s.length > start + 2) {
        result.push([start, s.length-1]);
    }

    return result;
};

let result = largeGroupPositions("abc");

console.assert(result.length === 0);

let expected = [[3,5],[6,9],[12,14]];
result = largeGroupPositions("abcdddeeeeaabbbcd");

console.assert(result.every((list, i) => list.every((value, j) => value === expected[i][j])));
